src/WebJobs.Script/Host/HostIdValidator.cs

```diff
@@ -0,0 +1,198 @@
+                    _validationScheduled = true;
+                }
+            }
+        }
+
+        internal async Task ValidateHostIdUsageAsync(string hostId)
+        {
+            try
+            {
+                if (!_storageProvider.ConnectionExists(ConnectionStringNames.Storage))
+                {
+                    return;
+                }
+
+                var primaryHostStateProvider = _serviceProvider.GetScriptHostServiceOrNull<IPrimaryHostStateProvider>();
+                if (primaryHostStateProvider != null && !primaryHostStateProvider.IsPrimary)
+                {
+                    // as an optimization, only validate on the primary host
+                    return;
+                }
+
+                HostIdInfo hostIdInfo = await ReadHostIdInfoAsync(hostId);
+
+                if (hostIdInfo != null)
+                {
+                    // an existing record exists for this host ID
+                    // verify the host name is the same as our host name
+                    if (string.Compare(_hostNameProvider.Value, hostIdInfo.Hostname, StringComparison.OrdinalIgnoreCase) != 0)
+                    {
+                        HandleCollision(hostId);
+                    }
+                }
+                else
+                {
+                    // no existing record, so write one, claiming this host ID for this host name
+                    // in this storage account
+                    hostIdInfo = new HostIdInfo
+                    {
+                        Hostname = _hostNameProvider.Value
+                    };
+                    await WriteHostIdAsync(hostId, hostIdInfo);
+                }
+            }
+            catch (Exception ex)
+            {
+                // best effort - log error and continue
+                _logger.LogError(ex, "Error validating host ID usage.");
+            }
+        }
+
+        private void HandleCollision(string hostId)
+        {
+            // see if the user has specified a level, otherwise default
+            LogLevel level = DefaultLevel;
+            string value = _environment.GetEnvironmentVariable(EnvironmentSettingNames.FunctionsHostIdCheckLevel);
+            if (value != null && Enum.TryParse<LogLevel>(value, out LogLevel parsedLevel))
+            {
+                level = parsedLevel;
+            }
+
+            string message = string.Format(Resources.HostIdCollisionFormat, hostId);
+            if (level == LogLevel.Error)
+            {
+                _logger.LogError(message);
+                _applicationLifetime.StopApplication();
+            }
+            else
+            {
+                // we only allow Warning/Error levels to be specified, so anything other than
+                // Error is treated as warning
+                _logger.LogWarning(message);
+            }
+        }
+
+        internal async Task WriteHostIdAsync(string hostId, HostIdInfo hostIdInfo)
+        {
```

Comment:
General question -- should we put a few more debug logs throughout here so we can track this should we need to? Examples -- when we write the blob, when we successfully validate? Mostly to ensure this is running as expected.

Reply:
I added a debug log below when we write the host ID record. Not sure we want to log when we successfully validate - that might just be noise?